cli/flox/src/commands/gc.rs

```diff
@@ -1,23 +1,291 @@
-        fields(progress = "Garbage collecting unused environment data")
-    )]
+    #[instrument(skip_all, fields(progress = "Collecting garbage"))]
     pub fn handle(self, flox: Flox) -> Result<()> {
-        env_registry::garbage_collect(&flox)?;
+        {
+            let span = info_span!(
+                "prune_environments",
+                progress = "Garbage collecting stale environments"
+            );
+            let _ = span.enter();
+            env_registry::garbage_collect(&flox)?;
+        }
+        run_store_gc()?;
 
         message::updated("Garbage collection complete");
         Ok(())
     }
 }
+
+/// Represents the stages of garbage collection in the logs of the
+/// `nix store gc` command.
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum GcProgress {
+    /// The initial state, which is very short lived.
+    Init,
+    /// Finding garbage collector roots.
+    FindingRoots,
+    /// If any stale links are found, this stage deletes them. If no stale links
+    /// are found, this stage may be skipped.
+    RemovingLinks,
+    /// Scanning store paths for liveness.
+    Scanning,
+    /// Deleting the stale store paths. Contains a message indicating which
+    /// store path is being deleted.
+    DeletingStorePaths(String),
+    /// We're done deleting things and shouldn't transition to any other state.
+    /// Contains a message indicating how many store paths were deleted and
+    /// how much space was reclaimed.
+    Done(String),
+}
+
+impl GcProgress {
+    fn new() -> Self {
+        GcProgress::Init
+    }
+
+    fn msg(&self) -> Option<String> {
+        use GcProgress::*;
+        match self {
+            Init => None,
+            FindingRoots => Some("Finding garbage collector roots".to_string()),
+            RemovingLinks => Some("Removing stale links".to_string()),
+            Scanning => Some("Scanning packages for liveness".to_string()),
+            DeletingStorePaths(msg) => Some(msg.clone()),
+            Done(msg) => Some(msg.clone()),
+        }
+    }
+
+    fn state_name(&self) -> &'static str {
+        use GcProgress::*;
+        match self {
+            Init => "init",
+            FindingRoots => "finding_roots",
+            RemovingLinks => "removing_links",
+            Scanning => "scanning",
+            DeletingStorePaths(_) => "deleting",
+            Done(_) => "done",
+        }
+    }
+
+    fn is_removing_links_line(line: &str) -> bool {
+        line.starts_with("removing stale link")
+    }
+
+    fn is_finding_roots_line(line: &str) -> bool {
+        line.starts_with("finding garbage collector roots")
+    }
+
+    fn is_scanning_line(line: &str) -> bool {
+        line.starts_with("cannot delete")
+    }
+
+    fn is_deletion_line(line: &str) -> bool {
+        line.starts_with("deleting '/nix/store")
+    }
+
+    fn is_done_line(line: &str) -> bool {
+        line.ends_with("MiB Freed")
+    }
+
+    fn new_progress_state(&self, line: &str) -> Option<GcProgress> {
```

Comment:
This is where we determine whether we're transitioning from one state to another.